add nested decorator and helper capabilities

dmanto opened the pull request:
Please consider this PR as a way to add nested decorators and / or helpers. It is based on a discussion that take place on Matrix's #mojo Room.

A nested helper would be added like this:

```typescript
app.addHelper('debug.foo.bar', (ctx, str) => {
  ctx.app.log.debug(str);
});
```
Then you will be able to call it in your context, like:
```typescript
app.get('/', async ctx => {
  ctx.debug.foo.bar('home route was just called');
  await ctx.render({text: 'Hello World!'});
});
```

kraih commented on src/app.ts:
```diff
@@ -270,17 +276,29 @@ export class App {
+      );
     } else {
-      this._contextClass.prototype[name] = fn;
+      target[methodName] = nestedNames.length == 0 ? fn : fn.bind(this);
```
Doesn't this bind to the app instance? And the helpers/decorators in the tests only work by accident because `app.config` exists too?

kraih replied:
A test using `ctx.stash` might be safer to ensure it really works.

dmanto replied:
- Yes, it binds to the app instance, other properties like ctx.stash cannot be accessed.
- I added a test using ctx.stash (it fails so far)

kraih commented on src/app.ts:
```diff
@@ -236,6 +236,12 @@ export class App {
    * });
+   *
+   * // Render response with header using nested helper
+   * app.addHelper('renderWith.Header', async (ctx, ...args) => {
```
Upper case method name.

kraih commented on src/app.ts:
```diff
@@ -270,17 +276,29 @@ export class App {
    * app.decorateContext('helloWorld', {get: () => 'Hello World!'});
+   *
+   * // Decorate context with nested getter
+   * app.decorateContext('hello.World', {get: () => 'Hello World!'});
```
Upper case getter/setter name.

dmanto replied:
changed to 'hello.world'

kraih commented:
Thanks. This looks pretty good overall, but i think the implementation might need proxy objects to pass `ctx` to the nested helper as `this`.

kraih commented:
And don't forget to squash your commits once you're done.

dmanto commented:
- The problem is that you cannot use the helpers without binding it, it has to be bound to current container (`ctx`) so the helper will be executed in the right context.
- This behavior can be observed in node repl, first logging `this` from a function at object level, then from a similar function but nested into the object hierarchy, like this:
```node
$>  node
Welcome to Node.js v19.7.0.
Type ".help" for more information.
> {stash: {alfa: 0}, nested: {bar: function() {console.log(this)}, other: 1}, foo: function() {console.log(this)}}.foo()
{
  stash: { alfa: 0 },
  nested: { bar: [Function: bar], other: 1 },
  foo: [Function: foo]
}
undefined
> {stash: {alfa: 0}, nested: {bar: function() {console.log(this)}, other: 1}, foo: function() {console.log(this)}}.nested.bar()
{ bar: [Function: bar], other: 1 }
undefined
> .exit
$>
```

- Nested helpers functions have to be bound to current context at Context creation, so those helpers can be correctly used during `ctx` lifespan.
- Nested helper functions are saved into the context hierarchy as they are added (through app.addHelper()), and then on context creation are bound and copied to the right location.
- TODO: Something similar should be done for getter and setters, the tests pass because they are using config helper which is available at app level.

About the proxy objects solution I guess you refer to a getter that returns the nested (last part) helper function. If this is the case I couldn't make it work, the problem is similar to the one was that `this` inside the helper function (named `nested.foo()` for example) refers to `ctx.nested` instead of `ctx`.

kraih commented:
Using [proxy objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) should sidestep the bind problem and result in much less code. I actually would have preferred two commits here, since the original broken solution would have been rather useful as a foundation for testing different approaches.

dmanto commented:
- I pushed a new revision, now implemented using a getter for the root level of the indent as you suggested.
- Nested setters and getters code was removed, only helpers are supported.
- I squashed all commits because de code is very different than the one in the previous version, so I think will be easer to review this way. Anyway I have many intermediate branches locally, please let me know if you prefer to have some particular commit in the middle.
- Added a test to check that the nested helper is not deeper than 2 levels (like ctx.foo.bar()), I put it into test/app.js because it seems to belong there better than in test/plugin-app.js

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,36 @@ export class App {
+      const [getterName, methodName] = nestedNames;
+      this._nestedFunctions[methodName] = fn;
+      const ctxScopedFunctions: any = {};
+      for (const key in this._nestedFunctions)
```
I try to only use `of` instead of `in` anymore.

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,36 @@ export class App {
+      });
+    else if (nestedNames.length == 2) {
+      const [getterName, methodName] = nestedNames;
+      this._nestedFunctions[methodName] = fn;
```
This approach seems broken, since it only works for a single getter name. Probably needs a test with at least two nested helpers like `ctx.foo.testHelper()` and `ctx.bar.testHelper()`.

dmanto replied:
fixed in last push

dmanto commented:
last push is incremental to yesterday's (just to facilitate comparison)

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,39 @@ export class App {
-      return fn(this, ...args);
-    });
+    const nestedNames = name.split('.');
+    if (nestedNames.length == 1)
```
Not using blocks for anything else than a one-liner looks horrible.

dmanto replied:
sure no problem

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,39 @@ export class App {
+      this._nestedFunctions[getterName] ??= {};
+      this._nestedFunctions[getterName][methodName] = fn;
+      const ctxScopedFunctions: any = {};
+      let that: unknown;
```
I bet this breaks when there is two or more active context instances, each keeping a copy of `const nested = ctx.nested;`, and then calling `nested.fooHelper();` at the same time. Then they would be using each other's stash and you get super hard to debug issues.

dmanto replied:
It actually breaks when you keep a copy of the getter.

For the record, you should be able to reproduce it with this simple Mojo app:

```javascript
/*
 * Minimal helper app to use to check nested helper concurrency
 */
import mojo from '../lib/core.js';

const app = mojo();

app.log.level = 'trace';

app.addHelper('nested.requestId', (ctx) => ctx.req.requestId);

app.get('/:kind', async ctx => {
    const {kind} = ctx.stash;
    const requestId = ctx.req.requestId;
    const nested = ctx.nested; // save nested getter
    const helperOriginalId = kind === 'safe' ? ctx.nested.requestId() : nested.requestId();
    await new Promise(r => setTimeout(r, 1000));
    const helperFinalId = kind === 'safe' ? ctx.nested.requestId() : nested.requestId();
    if (helperOriginalId === requestId && helperFinalId === requestId) {
        ctx.log.trace(`Ok, requestId = ${requestId} everywhere`);
        await ctx.render({ text: `Ok, requestId = ${requestId} everywhere` });
    } else {
        ctx.log.error(
            `Corrupted ctx: requesId: ${requestId} , helper Original: ${helperOriginalId}, helper Final: ${helperFinalId}`);
        await ctx.render({ text: `Not Ok, requesId: ${requestId} , helper Original: ${helperOriginalId}, helper Final: ${helperFinalId}`, status: 500 });
    }
});

app.start();

```
The idea is that a get on `'/safe'` doesn't use the saved getter base, but a get on  `/unsafe` or whatever different from `safe` uses it, and makes it fail.

So this works and logs all Ok traces:
```shell
$> autocannon -c 100 -d 10 -p 10 http://127.0.0.1:3000/safe
```
And this logs a lot of corrupted ctx errors:
```shell
$> autocannon -c 100 -d 10 -p 10 http://127.0.0.1:3000/unsafe
```

I will keep investigating this during the weekend and hopefully push a fix soon

kraih replied:
In the context class we already use `Proxy` in [two](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/src/context.ts#L66) [cases](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/src/context.ts#L198). If you need some inspiration.

dmanto replied:
Thanks, that helped, seems to be the way to go.
I will need some time to refactor several parts. In the meantime I added a test for the concurrency problem, and implemented a naive fix (just builds the functions object inside the getter every time its called)
Again, not squashed yet, to facilitate comparison

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,38 @@ export class App {
+      return this.decorateContext(getterName, {
+        get: function () {
+          const ctxScopedFunctions: any = {};
+          for (const [key, fn] of fnEntries)
```
Please no block-less loops and conditions. They are so ugly.

dmanto replied:
done

kraih commented on test/plugin-app.js:
```diff
@@ -8,13 +9,57 @@ t.test('Plugin app', async t => {
+      text: `Not Ok, requesId: ${requestId} , helper read: ${helperFinalId}`,
+      status: 500
+    });
+  });
```
This test case looks way too complicated and probably prone to race conditions.

kraih replied:
Should be possible to just make two [mock contexts](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/test/app.js#L933) and use those to trigger the problem with very little code.

dmanto replied:
done in last commit.
Previous to this one, there is a cache implemented inside the getter, so the resulting returned object is calculated only when the first get() happens.

About the Proxy implementation, I probably would need to proxy the ctx object itself, as according to the intents I was able to do so far it is the only way to actually receive the proper `this` (as target) in the get(target, prop, receiver) method from the proxy definition.

Is that the suggested approach? I mean to proxy the context object and not the ctx.nested object.

kraih replied:
I have not suggested a specific approach.

kraih commented on test/plugin-app.js:
```diff
@@ -113,6 +118,49 @@ t.test('Plugin app', async t => {
+  await t.test('Nested helper with wrong root name', async () => {
+    const logLevel = app.log.level;
+    app.log.level = 'fatal';
+    (await ua.getOk('/nested/innexistent_helper_with_stash/works'))
```
"innexistent" is not a word.

dmanto replied:
fixed

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,41 @@ export class App {
+      return this.decorateContext(getterName, {
+        get: function () {
+          (this as MojoContext)._cachedFnEntries ??= {};
+          if (!(this as MojoContext)._cachedFnEntries[getterName]) {
```
I prefer explicit `=== undefined` conditions. Then there's no guessing about type casts, which makes code much easier to read.

dmanto replied:
fixed

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,41 @@ export class App {
-      return fn(this, ...args);
-    });
+    const nestedNames = name.split('.');
+    if (nestedNames.length == 1) {
```
Use `===` instead of `==` whenever possible.

kraih commented on src/app.ts:
```diff
@@ -185,6 +185,7 @@ export class App {
   validator = new Validator();
 
   _contextClass: any = class extends ContextWrapper {};
+  _nestedFunctions: any = {};
```
This can probably be a bit better than `any`.

dmanto replied:
fixed, like Record<string, Record<string, MojoAction>>

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,41 @@ export class App {
+      return this.decorateContext(name, function (this: MojoContext, ...args: any[]) {
+        return fn(this, ...args);
+      });
+    } else if (nestedNames.length == 2) {
```
Same here about `===` instead of `==`.

dmanto replied:
fixed

kraih commented on test/plugin-app.js:
```diff
@@ -43,6 +47,7 @@ t.test('Plugin app', async t => {
     .name('mix');
 
   const ua = await app.newTestUserAgent({tap: t});
+  const ua2 = await app.newTestUserAgent({tap: t}); // second ua for concurrency tests
```
That comment doesn't really add much value.

kraih replied:
Looks like `ua2` is dead code.

kraih commented:
I think the current implementation works as expected, it's just a bit inefficient. We might go ahead and merge it, since it can be replaced with something more efficient later on. Most important is good test coverage for that.

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,41 @@ export class App {
+      this._nestedFunctions[getterName][methodName] = fn;
+      const fnEntries = Object.entries<MojoAction>(this._nestedFunctions[getterName]);
+      return this.decorateContext(getterName, {
+        get: function () {
```
Is it not possible to do `get: function (this: MojoContext) {` here and skip all the individual type assignments, or do i remember wrong?

dmanto replied:
fixed

kraih commented:
This is probably the time to squash the commits.

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,41 @@ export class App {
+      const fnEntries = Object.entries<MojoAction>(this._nestedFunctions[getterName]);
+      return this.decorateContext(getterName, {
+        get: function (this: MojoContext) {
+          this._cachedFnEntries ??= {};
```
Is there a reason this is created dynamically and not at the top?

dmanto replied:
Moved into Context class, please check.
Also changed the name for _cachedFunctions.

kraih commented on src/context.ts:
```diff
@@ -126,6 +126,7 @@ class Context extends EventEmitter {
    */
   stash: Record<string, any> = {};
 
+  _cachedFunctions: Record<string, Record<string, MojoAction>> = {};
```
These private properties should probably reflect that they are for nested helpers specifically.

kraih replied:
Like `_nestedHelperCache` or so.

kraih commented on src/app.ts:
```diff
@@ -185,6 +185,7 @@ export class App {
   validator = new Validator();
 
   _contextClass: any = class extends ContextWrapper {};
+  _nestedFunctions: Record<string, Record<string, MojoAction>> = {};
```
`_nestedHelpers`?

dmanto replied:
fixed using _nestedHelpersCache

kraih replied:
No, i mean `_nestedHelpers` would be a more appropriate name here.

kraih replied:
Nooooo, i meant here! `_nestedFunctions` should be `_nestedHelpers`.

dmanto replied:
should be ok now, _nestedHelpers is the prop in the app object, and _nestedHelpersCache the one in the ctx

kraih commented on src/app.ts:
```diff
@@ -236,11 +237,40 @@ export class App {
+          if (this._cachedFunctions[getterName] === undefined) {
+            const ctxScopedFunctions: Record<string, MojoAction> = {};
+            for (const [key, fn] of fnEntries) {
+              ctxScopedFunctions[key] = (...args: any[]) => fn(this as MojoContext, ...args);
```
Is the `this as MojoContext` still required here?

dmanto replied:
fixed.

dmanto commented:
rebased and squashed.
Anyway I will keep a non-squashed local branch for a few days, just in case

kraih commented on src/context.ts:
```diff
@@ -126,6 +126,7 @@ class Context extends EventEmitter {
    */
   stash: Record<string, any> = {};
 
+  _nestedHelpersCache: Record<string, Record<string, MojoAction>> = {};
```
Not ordered alphabetically.

dmanto replied:
fixed

marcusramberg reviewed (approved):
This looks like a good improvement to me. 👍🏻
